src/parser.rs

```diff
@@ -0,0 +1,555 @@
+use std::error::Error;
+
+use crate::command::{cmd, runnable};
+use crate::errors::SyntaxError;
+use crate::pipeline::Pipeline;
+use crate::redirect::{
+    InputRedirect, OutputRedirect, OutputRedirectAppend, Redirect, RedirectType,
+};
+use crate::sequence::{AndSequence, Sequence};
+use crate::token::Token;
+use crate::traits::{Runnable, ShellCommand};
+
+// Type aliases for readability
+type ShellCommandBox = Box<dyn ShellCommand>;
+type RunnableBox = Box<dyn Runnable>;
+type ShellCommandBoxes = Vec<ShellCommandBox>;
+type RunnableBoxes = Vec<RunnableBox>;
+
+pub fn tokenize(input: String) -> Result<RunnableBox, Box<dyn Error>> {
+    let input = preprocess(input);
+    let tokens = lex(input)?;
+    debug!("Tokens: {:?}", tokens);
+    let commands = parse(tokens)?;
+    Ok(commands)
+}
+
+pub fn preprocess(input: String) -> String {
+    // Remove leading and trailing whitespace
+    let mut input = input.trim().to_string();
+
+    // Remove trailing newline
+    if input.ends_with('\n') {
+        input.pop();
+    }
+
+    // Add semicolon to trigger final command
+    if !input.ends_with(';') {
+        input.push(';');
+    }
+
+    input
+}
+
+pub fn lex(input: String) -> Result<Vec<Token>, Box<dyn Error>> {
+    let mut escaped = false;
+    let mut quote_stack = Vec::<char>::new();
+    let mut current_token = String::new();
+    let mut tokens = Vec::<Token>::new();
+
+    for c in input.chars() {
```

Comment:
What we did in CL was that we had a "Scanner" which provided a `consume` and `peek` function, which makes the tokenising easier than looping through the characters like this.
Rather than looping over each character, each iteration of the main loop will basically add a token to the tokens.

Reply:
Yeah that's good. Let me try that.